galley/pkg/config/analysis/analyzers/auth/mtls.go

```diff
@@ -0,0 +1,185 @@
+//
+// The analyzer tries to act more generally by imagining service-to-service
+// traffic and detecting whether or not there's a conflict with regards to mTLS
+// policy. This means it will also detect explicit misconfigurations as well.
+//
+// Note this is very similar to `istioctl authn tls-check`; however this
+// inspection is all done via analyzing configuration rather than requiring a
+// connection to pilot.
+type MTLSAnalyzer struct{}
+
+// Compile-time check that this Analyzer correctly implements the interface
+var _ analysis.Analyzer = &MTLSAnalyzer{}
+
+// Metadata implements Analyzer
+func (s *MTLSAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "auth.MTLSAnalyzer",
+
+		// Each analyzer should register the collections that it needs to use as input.
+		Inputs: collection.Names{
+			metadata.K8SCoreV1Namespaces,
+			metadata.IstioAuthenticationV1Alpha1Meshpolicies,
+			metadata.IstioAuthenticationV1Alpha1Policies,
+			metadata.IstioNetworkingV1Alpha3Destinationrules,
+		},
+	}
+}
+
+// Analyze implements Analyzer
+func (s *MTLSAnalyzer) Analyze(c analysis.Context) {
+	// TODO Reuse pilot logic as a library rather than reproducing its logic
+	// here.
+
+	// While we visit every item, collect the set of namespaces that exist. Note
+	// that we will collect the namespace name for all resource types - this
+	// ensures our analyzer still behaves correctly even if namespaces are
+	// implicitly defined.
+	namespaces := make(map[string]struct{})
+
+	c.ForEach(metadata.K8SCoreV1Namespaces, func(r *resource.Entry) bool {
+		_, name := r.Metadata.Name.InterpretAsNamespaceAndName()
+		namespaces[name] = struct{}{}
+		return true
+	})
+
+	pc := mtls.NewPolicyChecker()
+	c.ForEach(metadata.IstioAuthenticationV1Alpha1Meshpolicies, func(r *resource.Entry) bool {
+		ns, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+		namespaces[ns] = struct{}{}
+
+		pc.AddMeshPolicy(r, r.Item.(*v1alpha1.Policy))
+		return true
+	})
+
+	c.ForEach(metadata.IstioAuthenticationV1Alpha1Policies, func(r *resource.Entry) bool {
+		ns, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+		namespaces[ns] = struct{}{}
+
+		err := pc.AddPolicy(r, r.Item.(*v1alpha1.Policy))
+		if err != nil {
+			c.Report(metadata.IstioAuthenticationV1Alpha1Meshpolicies, msg.NewInternalError(r, err.Error()))
+			return false
+		}
+		return true
+	})
+
+	drc := mtls.NewDestinationRuleChecker()
+	c.ForEach(metadata.IstioNetworkingV1Alpha3Destinationrules, func(r *resource.Entry) bool {
+		ns, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+		namespaces[ns] = struct{}{}
+
+		drc.AddDestinationRule(r, r.Item.(*v1alpha3.DestinationRule))
+		return true
+	})
+
+	// Here we explicitly handle the common case where a user specifies a
+	// MeshPolicy with no global MTLS rule
+	mpr := pc.MeshPolicy()
+	globalMtls, globalDR := drc.DoesNamespaceUseMTLSToService("istio-system", "istio-system", mtls.NewTargetService("*.local"))
```

Comment:
https://github.com/istio/istio/issues/18931 tracks allowing this. I think it's OK to leave that to a future PR, although it might not hurt to include a comment linking to the github issue so we're less likely to lose track.

Reply:
This is done.